Basic implementation of Crudella

smolijar opened the pull request.

smolijar committed:
Update tslint

smolijar committed:
Add basic implementation of the crud service

smolijar committed:
Add simple handler tests

smolijar committed:
Refactor methods in tests

smolijar committed:
Enhance hanlder result test to compaer more than id

smolijar committed:
Add authorization tests

smolijar committed:
Fix cs

smolijar committed:
Fix typo

smolijar committed:
Update test case message

smolijar committed:
Add tests for unimplemented default behaviour

smolijar committed:
Add repository implementation test

smolijar committed:
Add not found error handling

smolijar committed:
Remove redundant asyncs

smolijar committed:
Add options tests

smolijar committed:
Add prettier

smolijar committed:
Fix coding style using prettier

smolijar committed:
Add husky 🐶

smolijar committed:
Merge pull request #1 from AckeeCZ/feature/prettier

Feature/prettier

smolijar committed:
Use inline snapshots for scalar results

smolijar committed:
Refactor helpers

smolijar committed:
Add lint fix option, ci to test on husky

smolijar committed:
Rename get(handler) to detail(handler)

smolijar committed:
Refactor helpers, meta and bootstrap

smolijar committed:
Refactor handlers creation

smolijar committed:
Refactor root methods

smolijar committed:
Complete main exports

smolijar committed:
Fix docs path after refactor

smolijar committed:
Update changelog

smolijar committed:
Merge pull request #2 from AckeeCZ/refactor/default-implementation

Refactor/default implementation

smoliji commented on src/lib/context/crudContext.ts:
```diff
@@ -0,0 +1,69 @@
+import { Operation } from './operation';
+
+export interface BaseCrudContext<T, C> {
+    /** Http context */
```
Is there any reason to use block comment style for oneline comments on this place, and a single line comment for oneline on l.13?

smolijar replied:
@smoliji Aye, this is not just a block comment, but also a documentation magic :sparkles: comment.
I understand it look weird, but typedoc (or generally any documentation tool) generally skips normal comments and rightfully so.

smoliji commented on src/lib/crudella.ts:
```diff
@@ -0,0 +1,24 @@
+import { createHandlers } from './service/handlers';
+import { Definitions } from './settings/definitions';
+
+export const createService = <T extends { id: any }, C extends object>(defs: Definitions<T>) => {
```
I generally disagree with this requirement for the object with must-have id attribute, but it seems ok for the first version. Agree -> issue?

smolijar replied:
Agree, will create an issue. But I am afraid that the nature of Crudella inherently requires _an identifier_. It seems possible to have it defined as a custom template parameter.

smolijar replied:
#7

smoliji commented on src/lib/service/bootstrap.ts:
```diff
@@ -0,0 +1,33 @@
+import { CrudRepository } from 'lib/settings/repository';
+
+const createDefaultImplementation = <T, C>(): Omit<Required<Definitions<T>>, 'repository'> => ({
+    detail: () => Promise.reject(new Error('"detail" not implemented')),
```
It suprises me that a fan of FP like you does not use `constant` functions to represent these

smolijar replied:
There is no FP / utility library used in Crudella. Using custom `constant` or `always` is possible here, but since its just few lines in one block, I dare say it would not improve legibility (defining custom fuction for this fart). Only elegance somewhat.

smoliji commented on src/lib/service/handlers.ts:
```diff
@@ -0,0 +1,108 @@
+     * Fetch resource, throw error when resource missing.
+     * This method is used for handlers working with a single existing resource (get, update, delete)
+     */
+    const getSafe = (context: Pick<DetailContext<T, C>, 'id' | 'context' | 'options'>): Promise<T> =>
```
I'd change `Promise` - which accepts only genuine ES6 Promises to a more generic PromiseLike.

AFAIK Promise fails for e.g. Bluebird Promises (used by knex and most libraries)

smolijar replied:
Good point, did not know. I like that. Fixed.

smoliji commented on src/lib/service/handlers.ts:
```diff
@@ -0,0 +1,108 @@
+    const detailHandler = (options: any = {}) => async (id: number, context: C) => {
+        const dynamicOptions = await implementation.getOptions(Operation.DETAIL);
+        options = { ...dynamicOptions, ...options, ...(context as object) };
+        const ctx: DetailContext<T, C> = await getSafe({ id, context, options }).then(entity => ({
```
Please keep the same formatting of `.then` callings accross the project.

- `}).then(` on the same line here
- `).then(` on a newline on l.38 ...

smolijar replied:
It is consistent and automatically managed by prettier. If you dont like that, please suggest changes in https://github.com/AckeeCZ/prettier-config-ackee.

Notice that the blocks are not identical (even ignoreing line length, which generally are taken into account) -- this line is preceeded by `}` while 38 by `)`

smoliji commented on src/lib/service/handlers.ts:
```diff
@@ -0,0 +1,108 @@
+    const detailHandler = (options: any = {}) => async (id: number, context: C) => {
+        const dynamicOptions = await implementation.getOptions(Operation.DETAIL);
+        options = { ...dynamicOptions, ...options, ...(context as object) };
+        const ctx: DetailContext<T, C> = await getSafe({ id, context, options }).then(entity => ({
```
Is it necessary to combine await and then in the same function?

If you really want to use the function-style, I'd prefer an IIFE to do this:
```
ctx = await ((entity) => { ...})(await getSafe(...))
```
This should do just fine.

smolijar replied:
Agree it is distasteful. However this is even worse IMHO form the point of readability :crying_cat_face: 

Will refactor.

smolijar replied:
Fixed :heavy_check_mark:

smolijar committed:
Replace Promise with PromiseLike

smolijar committed:
Fix promiselike in getSafe

smolijar committed:
Refactor Promise.then in handlers

Snapshots are changed becasue due to whac typing, id was accidentely
passed to UpdateContext. Now its fixed and removed from snapshot

smolijar committed:
Refactor bootstrap options

smolijar committed:
Rename getSafe to safeDetail for the sake of consistency

smolijar committed:
Refactor contexts using context creators
